packages/caravan-wallets/src/jade.ts

```diff
@@ -0,0 +1,504 @@
+import { randomBytes } from "crypto";
+
+import {
+  BitcoinNetwork,
+  ExtendedPublicKey,
+  getPsbtVersionNumber,
+  PsbtV2,
+  MultisigAddressType,
+} from "@caravan/bitcoin";
+import {
+  Jade,
+  JadeInterface,
+  SerialTransport,
+  IJade,
+  IJadeInterface,
+  JadeTransport,
+  base64ToBytes,
+  bytesToBase64,
+  MultisigDescriptor,
+  SignerDescriptor,
+  ReceiveOptions,
+} from "jadets";
+
+import {
+  DirectKeystoreInteraction,
+  PENDING,
+  ACTIVE,
+  INFO,
+} from "./interaction";
+import { MultisigWalletConfig } from "./types";
+
+
+export const JADE = "jade";
+
+const DEFAULT_NETWORK = "mainnet";
+
+function variantFromAddressType(
+  t: MultisigAddressType,
+): MultisigDescriptor["variant"] {
+  switch (t) {
+    case "P2SH":
+      return "sh(multi(k))";
+    case "P2WSH":
+      return "wsh(multi(k))";
+    case "P2SH-P2WSH":
+      return "sh(wsh(multi(k)))";
+    default:
+      throw new Error(`Unsupported addressType ${t}`);
+  }
+}
+
+function fingerprintFromHex(xfp: string): Uint8Array {
+  return Uint8Array.from(Buffer.from(xfp, "hex"));
+}
+
+function parseBip32Path(path_i: string): number[] {
+  let path = path_i;
+  if (path.startsWith("m/")) {
+    path = path.substring(2);
+  } else if (path.startsWith("m")) {
+    path = path.substring(1);
+    if (path.startsWith("/")) {
+      path = path.substring(1);
+    }
+  }
+  const segments = path.split("/");
+  const result: number[] = [];
+  for (const segment of segments) {
+    let hardened = false;
+    let numStr = segment;
+    if (segment.endsWith("'") || segment.endsWith("h")) {
+      hardened = true;
+      numStr = segment.slice(0, -1);
+    }
+    const index = parseInt(numStr, 10);
+    if (isNaN(index)) {
+      throw new Error(`Invalid path segment: ${segment}`);
+    }
+    result.push(index + (hardened ? 0x80000000 : 0));
+  }
+  return result;
+}
+
+function extractPathSuffix(
+  fullPathStr: string,
+  baseDerivation: number[],
+): number[] {
+  const fullPath = parseBip32Path(fullPathStr);
+  if (
+    fullPath.length < baseDerivation.length ||
+    !baseDerivation.every((v, i) => v === fullPath[i])
+  ) {
+    throw new Error(
+      `Path "${fullPathStr}" does not extend base derivation [${baseDerivation.join(",")}]`,
+    );
+  }
+  return fullPath.slice(baseDerivation.length);
+}
+
+function walletConfigToDescriptor(
```

Comment:
Can we call this something like `walletConfigToJadeDescriptor`? it's not quite a descriptor but actually something specific to jade it looks like. For string encoded descriptors we do have a library that can actually encode from a config to a string descriptor (see [@caravan/descriptors](https://www.npmjs.com/package/@caravan/descriptors))